### ISSUE TITLE
[Feature] Retry after all tests

### ISSUE THREAD

**Issue by Diokuz** [microsoft/playwright#23462]
Hi there!

Right now, retries happens right after initial run. E.g.

```
test 1 failed
test 1 retry #1 failed
test 1 retry #2 passed
test 2 passed
test 3 passed
test 4 passed
...
test 523 passed
```

The problem here: if some backend (used in the test 1) is down for, say, 10 minutes, all retries will fail.

What if we could run retries in the end of the job? E.g.

```
test 1 failed
test 2 passed
test 3 passed
test 4 passed
...
test 523 passed
test 1 retry #1 failed
test 1 retry #2 passed
```

**Comment by Exoow** [comment]
Do they? In our CI logging it looks as if they are run at the end:

![image](https://github.com/microsoft/playwright/assets/18096566/7b719df3-f471-44cc-afaf-e13a633d8a78)

**Comment by mxschmitt** [comment]
> The problem here: if some backend (used in the test 1) is down for, say, 10 minutes, all retries will fail.

This sounds more like a problem which should solved somewhere else in your end-to-end testing infra. I'd suggest to find out why the backend is down and make it available instead! In the end its like more a good thing that it fails, maybe you should even add `--max-failures=20` so it fails already after 20 test failures to let you know instantly that your backend is down instead of wasting compute.

**Comment by Diokuz** [comment]
> I'd suggest to find out why the backend is down and make it available instead!

Makes sense! But we have like tens thousand tests and hundrends backends, and flaky endpoints is just our reality.

We dont want to know our backend is down. At least if it is down for less than 10 minutes. It is just our reality. There are special team, responsible for it, they always fixing it, but that takes some time.

> so it fails already after 20 test failures to let you know instantly that your backend is down instead of wasting compute.

20 failing runs is more wasting thing) Even than `setTimeout(10 minutes)`, because of smart CI cpu scheduler. But the most computing save approach here is long-retry. See

1. No retries. We have to retry CI job, and rerun all tests.
2. 2-3 retries. We have to retry CI job, and rerun all tests. Plus, some time will be wasted for retreis agains broken backends. 
3. 20+ retries. Tests will be running untill all backends are ok.
4. long retries (what I propose) takes minimum cpu and minimum job retries. Only one retry happens at the end of the job, and it will probably succeed, because in the most cases backend will be ok again.

What we want – is just long-retry cycle, to handle 10-minutes scale problems.

**Comment by Diokuz** [comment]
Another proposal here: retryTimeout. A time to wait before run test again.

**Comment by Vovencia** [comment]
You can use `postinstall` to replace one line in the file:

You need to replace `this._queue.unshift(result.newJob)` with `this._queue.push(result.newJob)`

Source file: https://github.com/microsoft/playwright/blob/release-1.40/packages/playwright/src/runner/dispatcher.ts#L126
Compiled file (in your project): `/node_modules/@playwright/test/lib/runner/dispatcher.js`

original answer: https://github.com/microsoft/playwright/issues/23354#issuecomment-1866525398

---

Можно с помощью `postinstall` заменить в файле одну строчку:

Нужно заменить `this._queue.unshift(result.newJob)` на `this._queue.push(result.newJob)`

Исходный файл: https://github.com/microsoft/playwright/blob/release-1.40/packages/playwright/src/runner/dispatcher.ts#L126
Скомпилированный файл (в вашем проекте): `/node_modules/@playwright/test/lib/runner/dispatcher.js`
